File: android/app/src/main/java/dev/bridgey/android/GamingFrameReassembler.kt
package dev.bridgey.android

/**
 * GAMING MODE POC. Sliding-window anti-replay check (same construction IPsec/DTLS use): accepts a
 * sequence number exactly once, tolerates UDP's out-of-order delivery within a bounded window, and
 * rejects both replays and packets too old to fit in the window. Sequence is treated as an unsigned
 * 32-bit counter (wraps after ~4 billion packets - session-lifetime, never a concern in practice).
 */
internal class ReplayWindow(private val windowSize: Int = 64) {
    private var highestSequence = -1L
    private var bitmask = 0L

    @Synchronized
    fun acceptAndRecord(sequence: Int): Boolean {
        val seq = sequence.toLong() and 0xFFFFFFFFL
        if (highestSequence < 0) {
            highestSequence = seq
            bitmask = 1L
            return true
        }
        val diff = seq - highestSequence
        if (diff > 0) {
            bitmask = if (diff >= windowSize) 1L else (bitmask shl diff.toInt()) or 1L
            highestSequence = seq
            return true
        }
        val backBits = -diff
        if (backBits >= windowSize) return false
        val bit = 1L shl backBits.toInt()
        if (bitmask and bit != 0L) return false
        bitmask = bitmask or bit
        return true
    }
}

/**
 * GAMING MODE POC (not wired into production decode). Reassembles MTU-safe UDP fragments into
 * complete video frame payloads, with the two loss behaviors the spec calls for:
 *   - a newer frame arriving preempts (drops) any older, still-incomplete frame immediately -
 *     never wait for retransmission of a stale realtime frame;
 *   - an incomplete frame that nothing newer ever supersedes is dropped after `staleFrameTimeoutMs`
 *     (handles "the stream paused" / "the last-ever fragment was the one lost" cases the preemption
 *     path can't catch on its own).
 * Either path invokes `onKeyframeOrConfigLost` if the dropped frame was flagged KEYFRAME or CONFIG,
 * so the caller can request a fresh one over the existing TCP control channel (reusing
 * VideoFrameType.KEYFRAME_REQUEST - no new recovery mechanism needed). Ordinary delta frame loss is
 * silent by design: dropping one frame and continuing is exactly the desired behavior, not an error.
 */
internal class GamingFrameReassembler(
    private val staleFrameTimeoutMs: Long = 200L,
    private val onKeyframeOrConfigLost: () -> Unit,
) {
    private class InProgress(fragmentCount: Int, val flags: Int, val firstSeenAtMs: Long) {
        val fragments = arrayOfNulls<ByteArray>(fragmentCount)
        var received = 0
    }

    private val inProgress = LinkedHashMap<Long, InProgress>()
    private var highestFrameIdSeen = -1L

    @Synchronized
    fun onFragment(
        frameId: Long,
        fragmentIndex: Int,
        fragmentCount: Int,
        flags: Int,
        nowMs: Long,
        payload: ByteArray,
    ): ByteArray? {
        if (frameId < highestFrameIdSeen) return null
        if (frameId > highestFrameIdSeen) {
            dropOlderThan(frameId)
            highestFrameIdSeen = frameId
        }
        val state = inProgress.getOrPut(frameId) { InProgress(fragmentCount, flags, nowMs) }
        if (fragmentIndex >= state.fragments.size) return null
        if (state.fragments[fragmentIndex] == null) {
            state.fragments[fragmentIndex] = payload
            state.received++
        }
        if (state.received != state.fragments.size) return null
        inProgress.remove(frameId)
        val total = state.fragments.sumOf { it?.size ?: 0 }
        val out = ByteArray(total)
        var offset = 0
        for (fragment in state.fragments) {
            fragment ?: return null
            fragment.copyInto(out, offset)
            offset += fragment.size
        }
        return out
    }

    @Synchronized
    fun dropStale(nowMs: Long) {
        val staleIds = inProgress.entries
            .filter { nowMs - it.value.firstSeenAtMs > staleFrameTimeoutMs }
            .map { it.key }
        for (id in staleIds) reportIfImportant(inProgress.remove(id))
    }

    private fun dropOlderThan(frameId: Long) {
        val staleIds = inProgress.keys.filter { it < frameId }
        for (id in staleIds) reportIfImportant(inProgress.remove(id))
    }

    private fun reportIfImportant(dropped: InProgress?) {
        if (dropped != null && (dropped.flags and (GamingUdpFraming.Flags.KEYFRAME or GamingUdpFraming.Flags.CONFIG)) != 0) {
            onKeyframeOrConfigLost()
        }
    }
}

File: macos/Sources/BridgeyMac/GamingFrameReassembler.swift
import Foundation

/// GAMING MODE POC. Swift mirror of ReplayWindow (Kotlin) - see that file for the construction
/// rationale (same sliding-window anti-replay IPsec/DTLS use).
final class ReplayWindow {
    private let windowSize: UInt64
    private var highestSequence: Int64 = -1
    private var bitmask: UInt64 = 0
    private let lock = NSLock()

    init(windowSize: Int = 64) {
        self.windowSize = UInt64(windowSize)
    }

    @discardableResult
    func acceptAndRecord(_ sequence: UInt32) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        let seq = Int64(sequence)
        if highestSequence < 0 {
            highestSequence = seq
            bitmask = 1
            return true
        }
        let diff = seq - highestSequence
        if diff > 0 {
            bitmask = UInt64(diff) >= windowSize ? 1 : (bitmask << UInt64(diff)) | 1
            highestSequence = seq
            return true
        }
        let backBits = UInt64(-diff)
        guard backBits < windowSize else { return false }
        let bit: UInt64 = 1 << backBits
        guard bitmask & bit == 0 else { return false }
        bitmask |= bit
        return true
    }
}

/// GAMING MODE POC. Swift mirror of GamingFrameReassembler.kt - see that file's doc comment for the
/// full loss-handling rationale (preempt-on-newer-frame, timeout-on-nothing-newer, keyframe/config
/// loss reported for TCP-side recovery, ordinary delta loss silent by design).
final class GamingFrameReassembler {
    private final class InProgress {
        var fragments: [Data?]
        var received = 0
        let flags: Int
        let firstSeenAtMs: Int64
        init(fragmentCount: Int, flags: Int, firstSeenAtMs: Int64) {
            fragments = Array(repeating: nil, count: fragmentCount)
            self.flags = flags
            self.firstSeenAtMs = firstSeenAtMs
        }
    }

    private var inProgress: [UInt64: InProgress] = [:]
    private var order: [UInt64] = []
    private var highestFrameIdSeen: Int64 = -1
    private let staleFrameTimeoutMs: Int64
    private let onKeyframeOrConfigLost: () -> Void
    private let lock = NSLock()

    init(staleFrameTimeoutMs: Int64 = 200, onKeyframeOrConfigLost: @escaping () -> Void) {
        self.staleFrameTimeoutMs = staleFrameTimeoutMs
        self.onKeyframeOrConfigLost = onKeyframeOrConfigLost
    }

    func onFragment(
        frameId: UInt64, fragmentIndex: Int, fragmentCount: Int, flags: Int, nowMs: Int64, payload: Data
    ) -> Data? {
        lock.lock()
        defer { lock.unlock() }
        let signedFrameId = Int64(frameId)
        if signedFrameId < highestFrameIdSeen { return nil }
        if signedFrameId > highestFrameIdSeen {
            dropOlderThan(frameId)
            highestFrameIdSeen = signedFrameId
        }
        let state: InProgress
        if let existing = inProgress[frameId] {
            state = existing
        } else {
            state = InProgress(fragmentCount: fragmentCount, flags: flags, firstSeenAtMs: nowMs)
            inProgress[frameId] = state
            order.append(frameId)
        }
        guard fragmentIndex < state.fragments.count else { return nil }
        if state.fragments[fragmentIndex] == nil {
            state.fragments[fragmentIndex] = payload
            state.received += 1
        }
        guard state.received == state.fragments.count else { return nil }
        inProgress.removeValue(forKey: frameId)
        order.removeAll { $0 == frameId }
        var out = Data()
        for fragment in state.fragments {
            guard let fragment else { return nil }
            out.append(fragment)
        }
        return out
    }

    func dropStale(nowMs: Int64) {
        lock.lock()
        defer { lock.unlock() }
        let staleIds = order.filter { id in
            guard let state = inProgress[id] else { return false }
            return nowMs - state.firstSeenAtMs > staleFrameTimeoutMs
        }
        for id in staleIds {
            let dropped = inProgress.removeValue(forKey: id)
            order.removeAll { $0 == id }
            reportIfImportant(dropped)
        }
    }

    private func dropOlderThan(_ frameId: UInt64) {
        let staleIds = order.filter { $0 < frameId }
        for id in staleIds {
            let dropped = inProgress.removeValue(forKey: id)
            order.removeAll { $0 == id }
            reportIfImportant(dropped)
        }
    }

    private func reportIfImportant(_ dropped: InProgress?) {
        guard let dropped, dropped.flags & (GamingUdpFraming.Flags.keyframe | GamingUdpFraming.Flags.config) != 0 else { return }
        onKeyframeOrConfigLost()
    }
}
